pkg/pfconfig/cpf/cpf.go

```diff
@@ -0,0 +1,128 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package cpf implements the JSON configuration for the Common Packet Forwarder.
+package cpf
+
+import (
+	"bytes"
+
+	"go.thethings.network/lorawan-stack/pkg/frequencyplans"
+	"go.thethings.network/lorawan-stack/pkg/pfconfig/shared"
+	"go.thethings.network/lorawan-stack/pkg/ttnpb"
+	"go.thethings.network/lorawan-stack/pkg/types"
+)
+
+type LorafwdGatewayConfig struct {
+	ID *types.EUI64
+}
+
+type LorafwdGWMPConfig struct {
+	Node            string
+	ServiceUplink   uint16
+	ServiceDownlink uint16
+}
+
+// LorafwdConfig represents the Lorafwd configuration of Semtech's UDP Packet Forwarder.
+type LorafwdConfig struct {
+	Gateway  LorafwdGatewayConfig
+	Filter   struct{}
+	Database struct{}
+	GWMP     LorafwdGWMPConfig
+	API      struct{}
+}
+
+func (conf LorafwdConfig) MarshalText() ([]byte, error) {
+	buf := &bytes.Buffer{}
+	if err := lorafwdTmpl.Execute(buf, conf); err != nil {
+		return nil, err
+	}
+	return buf.Bytes(), nil
+}
+
+// LoradGatewayConf contains the Lorad configuration for the gateway's server connection.
+type LoradGatewayConf struct {
+	BeaconEnable    bool        `json:"beacon_enable"`
+	BeaconPeriod    uint        `json:"beacon_period,omitempty"`
+	BeaconFreqHz    uint        `json:"beacon_freq_hz,omitempty"`
+	BeaconFreqNb    uint        `json:"beacon_freq_nb,omitempty"`
+	BeaconStep      uint        `json:"beacon_step,omitempty"`
+	BeaconDatarate  uint        `json:"beacon_datarate,omitempty"`
+	BeaconBwHz      uint        `json:"beacon_bw_hz,omitempty"`
+	BeaconPower     uint        `json:"beacon_power,omitempty"`
+	BeaconInfodesc  interface{} `json:"beacon_infodesc,omitempty"`
+	BeaconLatitude  float64     `json:"beacon_latitude,omitempty"`
+	BeaconLongitude float64     `json:"beacon_longitude,omitempty"`
+}
+
+type LoradSX1301Conf struct {
+	shared.SX1301Config
+	InsertionLoss     float32 `json:"insertion_loss"`
+	InsertionLossDesc string  `json:"insertion_loss_desc,omitempty"`
+	AntennaGainDesc   string  `json:"antenna_gain_desc,omitempty"`
+}
+
+// LoradConfig represents the Lorad configuration of Semtech's UDP Packet Forwarder.
+type LoradConfig struct {
+	SX1301Conf  LoradSX1301Conf  `json:"SX1301_conf"`
+	GatewayConf LoradGatewayConf `json:"gateway_conf"`
+}
+
+// BuildLorad builds Lorad configuration for the given gateway, using the given frequency plan store.
+func BuildLorad(gtw *ttnpb.Gateway, fps *frequencyplans.Store) (*LoradConfig, error) {
+	fp, err := fps.GetByID(gtw.FrequencyPlanID)
+	if err != nil {
+		return nil, err
+	}
+	sx1301Conf, err := shared.BuildSX1301Config(fp)
+	if err != nil {
+		return nil, err
+	}
+	var gatewayConf LoradGatewayConf
+	if len(gtw.Antennas) > 0 {
```

Comment:
Not sure if this makes sense - it seems that `SX1301` only allows for 1 antenna gain to be configured, what should happen when there are multiple?
Same question for location.

Reply:
IIRC, with the v1 and v1.5 ref design, there's usually only one antenna per concentrator and each concentrator gets one SX1301_Conf object.
@htdvisser can confirm.